fix unicode highlighting in commit searches

ijt opened the pull request:
Fixes #4512 

Test plan: unit test

After:
![Screen Shot 2019-06-25 at 17 02 41](https://user-images.githubusercontent.com/15530/60142656-9ba5a900-976f-11e9-82a6-fa7e36604a8f.png)

Before:
![Screen Shot 2019-06-25 at 17 07 52](https://user-images.githubusercontent.com/15530/60142641-86307f00-976f-11e9-89bc-898467cfd6e1.png)

keegancsmith commented on cmd/frontend/graphqlbackend/search_commits.go:
```diff
@@ -400,13 +401,24 @@ func displayRepoName(repoPath string) string {
 	const maxMatchesPerLine = 25 // arbitrary
 
+	// Make map from byte positions to character positions.
+	byteToChar := make(map[int]int)
```
can you switch this to a slice of ints, and preallocate the size?

ijt replied:
There are gaps in the domain so it doesn't make as much sense as a slice, but I'm changing it to preallocate the map.

keegancsmith commented on cmd/frontend/graphqlbackend/search_commits.go:
```diff
@@ -400,13 +401,24 @@ func displayRepoName(repoPath string) string {
 				line:      int32(i + 1),
-				character: int32(match[0]),
-				length:    int32(match[1] - match[0]),
+				character: int32(byteToChar[match[0]]),
```
are we guaranteed to have match align to unicode byte boundaries?

ijt replied:
I didn't find anything that says that, but the regexp library does something that looks very much like what you're describing: https://golang.org/src/regexp/regexp.go?s=23994:24039#L735. Also, the doc page says regexp expects UTF-8 encoded strings. https://golang.org/pkg/regexp/

Can you think of a case that might cause the regexp library to fail at this, even though it appears to do the right thing with multibyte characters?

keegancsmith commented on cmd/frontend/graphqlbackend/search_commits.go:
```diff
@@ -400,13 +401,24 @@ func displayRepoName(repoPath string) string {
+	var b, c int
+	bytes.Map(func(r rune) rune {
+		byteToChar[b] = c
+		b += utf8.RuneLen(r)
```
RuneLen can return `-1`

ijt replied:
Added something to handle it. `bytes.Map` sends a `utf8.RuneError` in that case.

keegancsmith commented on cmd/frontend/graphqlbackend/search_commits.go:
```diff
@@ -400,13 +401,24 @@ func displayRepoName(repoPath string) string {
+		c++
+		return r
+	}, data)
+	byteToChar[b] = c
```
I think you need to move this to inside of each line?

ijt replied:
Fixed.

keegancsmith commented:
This likely has poor perf since it will be O(n) per match, but a simple way to support this is via `utf8.RuneCount(data[:offset])`.

ijt commented:
> This likely has poor perf since it will be O(n) per match, but a simple way to support this is via `utf8.RuneCount(data[:offset])`.

I added a benchmark and switched to `utf8.RuneCount`. It's much tidier -- thanks for the suggestion. Here's how it compares to what was there before this PR:
```
benchmark                        old ns/op     new ns/op     delta
Benchmark_highlightMatches-4     3586          3612          +0.73%
```
Seems fine.

keegancsmith commented on cmd/frontend/graphqlbackend/search_commits.go:
```diff
@@ -401,19 +402,20 @@ func highlightMatches(pattern *regexp.Regexp, data []byte) *highlightedString {
 	var highlights []*highlightedRange
-	for i, line := range bytes.Split(data, []byte("\n")) {
-		for _, match := range pattern.FindAllIndex(bytes.ToLower(line), maxMatchesPerLine) {
+	for i, line := range bytes.Split(bytes.ToLower(data), []byte("\n")) {
```
unrelated to your change, but I'd suspect the bytes.ToLower is incorrect. If we want to ignore case, we should be making pattern case insensitive. Additionally, how does ToLower interact with unicode? Is it possible lowering a utf8 rune results in a shorter utf rune. That would give incorrect highlighting results.

ijt replied:
Good point. Filed https://github.com/sourcegraph/sourcegraph/issues/4791.

keegancsmith commented on cmd/frontend/graphqlbackend/search_commits.go:
```diff
@@ -401,19 +402,20 @@ func highlightMatches(pattern *regexp.Regexp, data []byte) *highlightedString {
-				character: int32(match[0]),
-				length:    int32(match[1] - match[0]),
+				character: int32(utf8.RuneCount(line[:match[0]])),
+				length:    int32(utf8.RuneCount(line[:match[1]]) - utf8.RuneCount(line[:match[0]])),
```
So this technically could be much faster, since this a `O(nk)` algorithm, where `n` is the number of matches and `k` is the length of the line. With just a little bit of book-keeping you can make it `O(k)`. However, this might be fine in practice, since we don't generally have very long lines. But you should probably benchmark a more complicated use case than your current usecase. Something like `pattern := "a"` and `data = bytes.Repeat([]byte{'a'}, 5000)`. Also validate what the longest line that can be returned. For example we have a max length in searcher. But in zoekt we might not?

ijt replied:
I updated the benchmark to highlight two long lines each with 5000 characters as you suggested. It shows the performance staying roughly the same with this change:

```
$ benchcmp old.txt runecount.txt 
benchmark                        old ns/op     new ns/op     delta
Benchmark_highlightMatches-4     81053         80321         -0.90%
```

ijt replied:
There's a hard coded `maxMatchesPerLine = 25` such that the time complexity is really `O(25 n) = O(n)`. If we later decide to remove that limit then it will make sense to add the bookkeeping you suggested.

keegancsmith commented on cmd/frontend/graphqlbackend/search_commits_test.go:
```diff
@@ -124,3 +127,132 @@ func TestExpandUsernamesToEmails(t *testing.T) {
+		wantErr bool
+	}{
+		{
+			name: "https://github.com/sourcegraph/sourcegraph/issues/4512 case 1",
```
these names are quite frustrating to match against and read in go test. Can you not give it a more succinct name. If you want to reference the issue for this, maybe just put that as a comment?

ijt replied:
Done.

ijt replied:
Done.
